feat: date picker revamp

karrui opened the pull request:
## Problem
<!-- What problem are you trying to solve? What issue does this close? -->
I'm sorry. 

This PR updates the date picker to be stop using `type=date`, since screen readers still struggle with that input type (mistaking the date, month, and year subinputs as steppers). Safari also forces their native date picker to be shown over our calendar, which is not what we want since the native date picker is not fully featured enough for our use cases (such as excluding specific dates, etc).

In addition, did a bunch of housekeeping such as separating Calendar and DatePicker/DateRangePicker components and themes. Also updates designs to look like what the designers intended, and use `Drawer` to display the datepicker in mobile views. 

Also closes #4726 

## Solution
<!-- How did you solve the problem? -->

**Breaking Changes** 
<!-- Does this PR contain any backward incompatible changes? If so, what are they and should there be special considerations for release? -->
- [x] No - this PR is backwards compatible, purely frontend change.

**Features**:

- Add new `DatePicker` and `DateRangePicker` components
- Show `Calendar` in a drawer in mobile viewports.

**Improvements**:

- Extract out `Calendar` component

## Before & After Screenshots
Storybook has new stories. If you are lazy, here are the screenshots!!!

<img width="958" alt="Screenshot 2022-09-14 at 2 48 48 PM" src="https://user-images.githubusercontent.com/22133008/190081119-d5971d76-375b-47c8-ae0f-199ec7594e89.png">
<img width="338" alt="Screenshot 2022-09-14 at 2 49 12 PM" src="https://user-images.githubusercontent.com/22133008/190081289-59f79fdf-ff0c-4419-b751-de4fab094f9a.png">
<img width="919" alt="Screenshot 2022-09-14 at 2 49 59 PM" src="https://user-images.githubusercontent.com/22133008/190081596-65e998dd-840b-4ea5-8978-7c300b5d8cfe.png">
<img width="448" alt="Screenshot 2022-09-14 at 2 50 23 PM" src="https://user-images.githubusercontent.com/22133008/190081673-bac52d34-a269-454b-903f-eafbb97c300f.png">


## Tests
<!-- What tests should be run to confirm functionality? -->
DateField tests has been updated.

## Deploy Notes
<!-- Notes regarding deployment of the contained body of work.  -->
<!-- These should note any new dependencies, new scripts, etc. -->

**New dependencies**:

- `"react-input-mask"` : masking for date field

karrui committed:
feat: add initial DatePicker component using text type input

karrui committed:
feat: add input mask and manual text entry handling

karrui committed:
feat: add theme for faux input for new date input

karrui committed:
feat: move and make Calendar its own prop

todo: make specific theme for calendar

karrui committed:
refactor: move Calendar into its own component

karrui committed:
feat: use new DatePicker2 component (to be renamed) in date field

also update stories and validation

karrui committed:
ref: move Calendar props into CalendarBase directory

so range calendar can reuse shared props

karrui committed:
feat: add RangeCalendar component to allow multi-date selection

karrui committed:
feat: add more DatePicker stories

karrui committed:
fix: update calendar date comparison for bg

karrui committed:
feat: export RangeCalendar from Calendar dir

karrui committed:
fix(DateField): correctly validate and set answer for submission

karrui committed:
feat: add initial DateRangePicker component

karrui committed:
feat: add close calendar handling and focus

better UX to go back to the input instead of the calendar button when date has been selected

karrui committed:
ref: extract business logic into useDatePicker hook

karrui committed:
feat: add MobileDatePicker component

karrui committed:
feat: add better styling to mobile date picker

karrui committed:
feat: render Mobile or Desktop variant of datepicker based on viewport

karrui committed:
feat: handle date range picker date parsing

karrui committed:
style: update width of date range input, enable mobile scroll

to feel like real input

karrui committed:
ref: move DateRangePicker to own component

karrui committed:
feat: use controlled state for popover

karrui committed:
feat: add click handler instead of popover trigger

karrui committed:
style: use custom popover close button component

karrui committed:
feat: use context to determine mobile vs desktop datepicker behaviour

karrui committed:
feat: do the same for DateRangePicker

karrui committed:
feat: replace usage of old date input with revamped components

karrui committed:
fix: update usage DatePicker in EditDate component

karrui committed:
fix: update usage of DateRangePicker

karrui committed:
feat: remove old DatePicker and let DatePicker2 take over

karrui committed:
feat: separate date picker and calendar themes

karrui committed:
Merge branch 'form-v2/develop' into form-v2/date-picker-revamp

# Conflicts:
#	frontend/src/components/DatePicker/DateInput.tsx
#	frontend/src/features/admin-form/create/builder-and-design/BuilderAndDesignDrawer/EditFieldDrawer/edit-fieldtype/EditDate/EditDate.tsx

karrui committed:
feat(Calendar): remove commented theme code

karrui committed:
ref: use parseDate function helper on validation

karrui committed:
feat(DatePicker): allow passing of monthsToDisplay prop to calendar

wanlingt commented:
are u ok

karrui committed:
fix: update myinfo preview values for new date format

karrui committed:
feat: add invalid, disabled, and readonly styling to datepicker

karrui committed:
fix: update date field validation

karrui committed:
test: update DateField tests

justynoh commented on frontend/src/components/Calendar/CalendarBase/types.ts:
```diff
@@ -0,0 +1,8 @@
+  'colorScheme' | 'isDateUnavailable' | 'monthsToDisplay'
+>
+
+export type DateRangeValue = [null, null] | [Date, null] | [Date, Date]
```
is there a reason for not using something like `{start?: Date, end?: Date}` instead? that seems easier to read than arrays of two values

karrui replied:
no reason, preference. if you also no preference i'll stick to this since it'll probably not be worth the effort to rejig to an object

justynoh commented on frontend/src/components/Calendar/CalendarBase/DayOfMonth.tsx:
```diff
@@ -64,7 +62,7 @@ export const DayOfMonth = forwardRef<DayOfMonthProps, 'button'>(
     )
 
-    const styles = useMultiStyleConfig(DATE_INPUT_THEME_KEY, {
+    const styles = useMultiStyleConfig('Calendar', {
```
should the calendar theme key be pulled into a constant?

karrui replied:
following chakra convention, not sure why they don't use constants too though lol. Probably nbd

karrui committed:
Merge branch 'form-v2/develop' into form-v2/date-picker-revamp

# Conflicts:
#	frontend/src/features/admin-form/create/builder-and-design/BuilderAndDesignDrawer/EditFieldDrawer/edit-fieldtype/EditDate/EditDate.tsx

karrui committed:
Merge branch 'form-v2/develop' into form-v2/date-picker-revamp

# Conflicts:
#	frontend/src/components/DatePicker/DateInput.tsx

justynoh commented on frontend/src/components/DatePicker/DatePickerContext.tsx:
```diff
@@ -0,0 +1,235 @@
+    defaultValue: defaultInputValue ?? formatInputValue(internalValue),
+    value: inputValue,
+    onChange: onInputValueChange,
+  })
```
For readability, should we annotate these two `useControllableStates` with their types, or add a comment explaining the difference between these two sets of internal values?

karrui replied:
19a6f2d

justynoh commented on frontend/src/components/DateRangePicker/DateRangePickerContext.tsx:
```diff
@@ -0,0 +1,321 @@
+
+  const handleUpdateInputs = useCallback(
+    (nextRange: DateRangeValue) => {
+      // Replace invalid dates with null
+      const sortedRange = nextRange.sort((a, b) =>
+        a && b ? a.getTime() - b.getTime() : 0,
+      )
+
+      // Replace invalid dates with null
+      const validRange = sortedRange.map((date) =>
+        isValid(date) ? date : null,
+      ) as DateRangeValue
```
The first comment here is probably wrong haha

justynoh commented on frontend/src/components/DateRangePicker/components/DateRangePickerInput.tsx:
```diff
@@ -0,0 +1,81 @@
+          mask="99/99/9999"
+          value={startInputDisplay}
+          onChange={handleStartDateChange}
+          placeholder={displayFormat.toLowerCase()}
+          maskPlaceholder={displayFormat.toLowerCase()}
```
nit: we could move this conversion up so that we don't have so many identical function calls as prop inputs

karrui replied:
[bd9e907](https://github.com/opengovsg/FormSG/pull/4855/commits/bd9e9077bfd48509c5c94cbc606fcf581a1da17f)

karrui committed:
ref: expose placeholder from context instead of displayFormat

since that's all the current display format is used for

justynoh reviewed (approved):
lgtm, just one point of clarification on the type of date ranges

karrui committed:
fix: add comments, correctly format date with given displayFormat

wanlingt commented:
wait ah im still reviewing

wanlingt reviewed (changes requested):
found a few bugs
1. calendar in the public form displays the day BEFORE the min date. get a form submission error when that date is submitted, but the user should not be allowed to select it anyway.
I looked into it, it seems that `isDateUnavailable` is being returned as `false` for the day before
e.g. i chose a valid date range of 15/09/2022 - 15/09/2022, but 14 and 15/09 both show up
<img width="1624" alt="Screenshot 2022-09-15 at 5 39 51 PM" src="https://user-images.githubusercontent.com/56983748/190948649-8cc2d46e-e990-4925-a9b6-51c45e19a86e.png">
<img width="1624" alt="Screenshot 2022-09-15 at 5 39 59 PM" src="https://user-images.githubusercontent.com/56983748/190948656-3e37484d-761f-40d3-b25d-8dbb34823ebf.png">

2. more of a nit: when the date is near the end of the page, if the calendar below the field it lengthens the page. should restrict it to opening above the field
<img width="1624" alt="Screenshot 2022-09-15 at 4 41 02 PM" src="https://user-images.githubusercontent.com/56983748/190948712-788d754b-0b48-49ab-b4ce-c3e4711a249a.png">

also a separate suggestion when I was looking at UnlockedResponses - should we filter the responses table view when we filter the responses by date? rn the filter only applies to the downloaded csv.

karrui commented:
argh, that's due to timezones

karrui committed:
feat: update date field to consistently be in UTC

karrui committed:
feat: disable dates in customMaxDate that is before customMinDate

does not work the other way round of course

karrui commented:
@wanlingt fixed 1, ignoring 2. Datepicker on top may have the opposite problem of showing white bg on the top anyways. No big deal to extend from bottom too.

wanlingt reviewed (approved):
lgtm!

karrui committed:
feat: make screen readers read out full day name

karrui committed:
feat: add aria-live region for multiple calendars

karrui committed:
feat: move CalendarAria to own component

speaks out selected date and ranges in addition to current displayed month

karrui committed:
feat: update aspect ratio of calendar dates

karrui committed:
feat: update calendar button aria to "Select from datepicker"

timotheeg commented:
@karrui does this PR needs more reviews?

karrui commented:
> @karrui does this PR needs more reviews?

fixing pearly's comments, you can review if you want to, haha!

karrui committed:
feat: update keyboard to be shown in mobile when inputting dates

karrui committed:
feat: add aria-live region to input to announce selected date

karrui committed:
fix: add styles to allow Calendar to render properly in drawer on iOS

Drawer bottom was getting cut off on iOS grrrr

karrui committed:
fix: prevent drawer from being cut off for shorter mobile heights

can also happen if drawer is displaying a bunch of months

karrui committed:
feat: update aria-label of datepicker button to "date picker"

voiceover was pronouncing datepicker (single word) as data-piker for some reason. Might be because it is not a real word.

karrui committed:
feat: refocus input when date picker calendar is closed

karrui committed:
feat: use browser's locale as day's aria-label instead of formatting

more general

karrui commented:
Reopening to retrigger PR

karrui commented:
i regret. the CI/CD PR has not been merged in yet...

karrui committed:
Merge branch 'develop' into form-v2/date-picker-revamp
